onnx/defs/nn/defs.cc

```diff
@@ -101,20 +101,12 @@ void convPoolShapeInference(
     if ((nullptr != auto_pad_attr) && (auto_pad_attr->s() != "VALID")) {
       int input_dims_size = static_cast<int>(n_input_dims);
       for (int i = 0; i < input_dims_size; ++i) {
-        int64_t residual = 0;
-        if (strides[i] > 1) {
-          if (!input_shape.dim(2 + i).has_dim_value()) {
-            continue;
-          }
-          residual =  input_shape.dim(2 + i).dim_value();
-          while (residual > 0) {
-            residual -= strides[i];
-          }
-        }
-        int64_t total_pad = residual == 0 ? kernel_shape[i] - strides[i] : kernel_shape[i] + residual;
-        if (total_pad < 0) {
-          fail_shape_inference("Stride is bigger than Kernel shape");
+        if (!input_shape.dim(2 + i).has_dim_value()) {
+          continue;
         }
+        int64_t dim_value = input_shape.dim(2 + i).dim_value();
+        int64_t legacy_target_size = (dim_value + strides[i] - 1) / strides[i];
+        int64_t total_pad = (legacy_target_size - 1) * strides[i] + kernel_shape[i] - dim_value;
```

Comment:
What if there is the corner case,
input shape: 4, kernel shape: 2, strides: 4
then legacy_target_size = 1, total_pad will be negative value -2.

Reply:
Thanks @stevenlix. Actually I had a similar concern when I first looked at the ORT shape computation logic. I feel this computation is more complex than it should be. If you substitute `legacy_target_size` from line 108 in the formula for `total_pad` in line 109, and after simplification, it is simply - `kernel_shape[i] - 1` (and this makes sense to me). Pads computation need not invlove strides at all as they are not related. Pads are to be computed only so that the filter is able to be fully accomodated for each pixel. So, I have made a change now. Please correct me if I am wrong.

Reply:
@linkerzhang - Please comment on this line - 

`int64_t total_pad = kernel_shape[i] - 1;`

Reply:
@hariharans29 I think this will result in different pad number, which changes the kernel center position and leads to different result.

Reply:
@BowenBao - Thank you! But thinking out loud naively here, let's forget about the corner case that @stevenlix  pointed out for this discussion. The current pad computation logic used by ORT is here - 
https://github.com/microsoft/onnxruntime/blob/a7039601c4bf469131569d0c1261f13cd6bd67d9/onnxruntime/core/providers/cpu/nn/conv_base.h#L39

This was the previous logic in this same PR - 
`int64_t legacy_target_size = (dim_value + strides[i] - 1) / strides[i];
int64_t total_pad = (legacy_target_size - 1) * strides[i] + kernel_shape[i] - dim_value;`

Please substitute `legacy_target_size` into `total_pad` and simplify by doing some simple math, the result will be (if my math is correct) `kernel_shape[i] - 1` (which is the new logic I proposed). [Here](https://stackoverflow.com/questions/37674306/what-is-the-difference-between-same-and-valid-padding-in-tf-nn-max-pool-of-t) is another detailed discussion about this and it aligns with the ` kernel_shape[i] - 1`

As long as the total pads computed is the same, it should not affect kernel center position.